### DISCUSSION TITLE
Lock zoom level

### DISCUSSION THREAD

**Discussion by demiantres** [mapsforge/vtm#1028] [Q&A] [answered]
Locus Maps has a useful feature for locking the zoom level. When locked then only raster tiles from a specific zoom level are used independent from the current map zoom level.

One option for the same effect in VTM is to implement a custom "UrlTileDataSource.query" (or "BitmapTileLoader.loadTile") method which stitches together a tile from multiple tiles or cuts out a region from an existing tile. But this is quite slow/complex.

Is there a better solution for a lock function?

**Comment by devemux86** [comment]
Displaying bitmap tiles from higher zoom levels would result in a blurry map.

All `TileSource` builders have methods to limit the zoom levels (see `DefaultSources`).
Could use a `OverzoomTileDataSource` like in vector maps (see `MapilionMvtTileSource`)?

**Reply by demiantres** [reply]
I have already tested an ``OverzoomTileDataSource`` but this won't work because it loads the wrong tiles. The problem here is that the tile image from the server does not actually match the tile to be shown in the map.

**Reply by devemux86** [reply]
So far it is being used without any problems in vector tile sources, like `MapilionMvtTileSource`.

**Comment by demiantres** [comment]
The attached video shows what happens when setting the overzoom to 14. Starting with zoom level 14 the tiles get messed up.

https://user-images.githubusercontent.com/5980561/229297469-e099143d-25ef-4763-917c-1cc877077cf3.mp4


Code:
```
// BitmapTileSource.java
// original:
new UrlTileDataSource(this, new TileDecoder(), getHttpEngine());

// modified:
return new OverzoomTileDataSource(new UrlTileDataSource(this, new TileDecoder(), getHttpEngine()), 14);
```

**Comment by devemux86** [comment, marked as answer]
Overzoom was created mainly for vector maps where rendering at large zoom can be done with content from upper levels.

Bitmap maps display images. Setting `zoomMax` in their builder, they show zoomed / blurry the tiles from upper zoom levels.

```java
public static Builder<?> OPENSTREETMAP = BitmapTileSource.builder()
        .url("https://tile.openstreetmap.org")
        .zoomMax(14);
```

![VTM](https://user-images.githubusercontent.com/3484020/229300991-a7f388c2-f150-4d43-ad0f-23b0f73866d0.png)

**Reply by demiantres** [reply]
Got it. This is exactly what I was looking for. Thank you.
